doc/neps/nep-0042-new-dtypes.rst

```diff
@@ -0,0 +1,796 @@
+some conbinations should cast safely or preserve the Unit (both of which the
+"base" ``Float64`` would discard).
+In the proposed implementation this is not possible, since the two implementations
+are not aware of each other.
+
+
+**Notes:**
+
+The proposed ``CastingImpl`` this designed to be compatible with the
+``UFuncImpl`` proposed in NEP YY.
+While initially it will be a distinct object, the aim is that ``CastingImpl``
+can be a subclass of ``UFuncImpl``.
+Once this happens, this will naturally allow the use of a ``CastingImpl`` to
+pass around a specialized casting function directly if so wished.
+
+In the future, we may considering adding a way to spell out that specific
+casts are known to be *not* possible.
+
+
+C-Side API
+^^^^^^^^^^
+
+.. note:: At the time of writing, this API is a general design goal.
+          Due to the size of the proposed changes, details and names will be
+          in flux and updated as necessary.
+
+DType creation
+""""""""""""""
+
+As already mentioned in NEP 41, the interface to define new DataTypes in C
+is modelled after the limited API in Python, the above mentioned slots,
+and some additional necessary information will thus be passed within a slots
+struct and identified by ``ssize_t`` integers::
+
+    static struct PyArrayMethodDef slots[] = {
+        {NPY_dt_method, method_implementation},
+        ...,
+        {0, NULL}
+    }
+
+    typedef struct{
+      PyTypeObject *typeobj;    /* type of python scalar */
+      int flexible;             /* Is the dtype flexible? */
+      int abstract;             /* Is the dtype abstract? */
+      int flags;                /* Currently only the "needs API" flag */
+      /* NULL terminated CastingImpl; is copied and references are stolen */
+      CastingImpl *castingimpls[];
+      PyType_Slot *slots;
+    } PyArrayDTypeMeta_Spec;
+
+    PyObject* PyArray_InitDTypeMetaFromSpec(
+            PyArray_DTypeMeta *user_dtype, PyArrayDTypeMeta_Spec *dtype_spec);
+
+all of this information will be copied during instantiation.
+
+The proposed method slots are (prepended with ``NPY_dt_``), these are
+detailed above and given here for summary:
+
+* ``is_native(self) -> {0, 1}``
+* ``ensure_native(self) -> dtype``
+* ``default_descr(self) -> dtype`` (return must be native and should normally be a singleton)
+* ``get_sort_function(self, NPY_SORTKIND sort_kind, void **out_sortfunction) -> success {-1, 0}``.
+  ``out_sortfunction`` may be ``NULL`` if the sort kind is unknown or not implemented.
+* ``setitem(self, char *item_ptr, PyObject *value) -> {-1, 0}``
+* ``getitem(self, char *item_ptr, PyObject (base_obj) -> object or NULL``
+* ``discover_descr_from_pyobject(cls, PyObject) -> dtype or NULL``
+* ``common_dtype(cls, other) -> DType, NotImplemented, or NULL``
+* ``common_instance(self, other) -> dtype or NULL``
+
+If not set, most slots are filled with slots which either error or defer automatically.
```

Comment:
Often yes... But I think the issue is for container types (covariance vs. contravariance). It all confuses me a lot. The main thing is that it is never really *safe* to cast to the super class, but is it always *unsafe*? If I cast from Float64 to UnitFloat64 and it subclasses, should we allow that as "unsafe". Is `nonzero` still correct if I derive an NAInt64 from Int64?

Wikipedia says: Mutable data types which act as both sources and sinks should be invariant (https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)), but to be honest I have not fully grasped the theory. But whenever I tried to think about a good generalization something seemed to make it complicated...

Reply:
However, I think it might be fine on AbstractDTypes, and even inheriting "defaults" from the AbstractDType, even for slots/casting! You just must not subclass *concrete* dtypes. I am just thinking we can defer that a bit for now, probably.

Reply:
Hmm, thinking more, I'm now wondering about not being allowed to subclass `float64` - and it may again be good to be super-explicit: I think from what you write as an dtype instance as something that completely describes what some pieces of memory mean - this to me makes sense (though would be good to be explicit how it goes for units like in datetime!). But what about the dtype above? Is it that the class knows what the meaning is, and the instance only defines possible details of the storage? But why would that need to prevent subclassing?

Though probably none of this matters too much - there is real clarity about what the `AbstractDType` is, which I think is the more important distinction.

Reply:
I am cautious about subclassing, and still think it is a bad idea. We can make it easy to wrap or fallback to other datatypes, but...
There is actually a *real* and valid use case for subclassing which for sure does not violate Liskov's substitution principle: In theory (not sure how realistic) CuPy, could subclass float64, to add information that they need to work on the GPU, but still share implementation (and even be compatible with) NumPy.
(adding a sentence for that, may get a bit much, but it is an important possibility I think)